File: ios/Talise/Features/Profile/ClaimHandleSheet.swift
import SwiftUI

/// Bottom sheet that lets the user claim a `<name>.talise.sui` subname
/// after onboarding (or if the silent KYC auto-claim failed).
///
/// Flow:
///   1. Debounced GET /api/username/check?u=<input> on every keystroke
///   2. Tap "Claim" → POST /api/username/claim — operator wallet pays
///      gas + signs the SuiNS mint, user pays nothing
///   3. On success, refresh AppSession so HomeView/Profile pick up the
///      new handle from /api/me
struct ClaimHandleSheet: View {
    @Environment(AppSession.self) private var session
    @Environment(\.dismiss) private var dismiss

    @State private var input: String = ""
    @State private var availability: AvailabilityState = .empty
    @State private var checkTask: Task<Void, Never>?
    @State private var claiming = false
    @State private var error: String?
    @State private var claimed: String?
    @FocusState private var focused: Bool

    enum AvailabilityState: Equatable {
        case empty
        case checking
        case available
        case taken
        case reserved
        case invalid
        case rpcError
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            if let claimed {
                successView(handle: claimed)
            } else {
                form
            }
        }
        .background(TaliseColor.bg.ignoresSafeArea())
        .presentationDragIndicator(.visible)
        .onAppear {
            if input.isEmpty, case .ready(let user) = session.phase {
                input = user.suggestedHandle()
                scheduleCheck(input)
                focused = true
            }
        }
    }

    // MARK: - Form

    private var form: some View {
        VStack(alignment: .leading, spacing: 22) {
            VStack(alignment: .leading, spacing: 6) {
                MicroLabel(text: "Claim your name", color: TaliseColor.fgDim).kerning(1.5)
                Text("Pick your Talise handle")
                    .font(TaliseFont.heading(24, weight: .medium))
                    .kerning(-0.8)
                    .foregroundStyle(TaliseColor.fg)
                Text("People send to you with name@talise.sui — easier to share than a 0x address.")
                    .font(TaliseFont.body(13, weight: .light))
                    .foregroundStyle(TaliseColor.fgMuted)
                    .padding(.top, 2)
            }

            handleInput
            statusRow

            if let error {
                MicroLabel(text: error, color: TaliseColor.danger)
            }

            claimButton

            Spacer(minLength: 40)
        }
        .padding(.horizontal, 24)
        .padding(.top, 20)
    }

    private var handleInput: some View {
        HStack(spacing: 0) {
            TextField("alice", text: $input)
                .focused($focused)
                .autocorrectionDisabled()
                .textInputAutocapitalization(.never)
                .textContentType(.username)
                .keyboardType(.asciiCapable)
                .font(TaliseFont.heading(20, weight: .medium))
                .kerning(-0.4)
                .foregroundStyle(TaliseColor.fg)
                .tint(TaliseColor.accent)
                .onChange(of: input) { _, new in
                    let cleaned = sanitize(new)
                    if cleaned != new { input = cleaned }
                    scheduleCheck(cleaned)
                }

            Text("@talise.sui")
                .font(TaliseFont.body(15, weight: .light))
                .foregroundStyle(TaliseColor.fgMuted)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 14)
        .background(TaliseColor.usernameCard)
        .clipShape(RoundedRectangle(cornerRadius: 16))
    }

    private var statusRow: some View {
        HStack(spacing: 8) {
            statusIcon
            Text(statusText)
                .font(TaliseFont.body(12, weight: .light))
                .foregroundStyle(statusColor)
        }
        .frame(height: 18)
        .padding(.horizontal, 4)
    }

    @ViewBuilder
    private var statusIcon: some View {
        switch availability {
        case .checking:
            ProgressView().controlSize(.mini).tint(TaliseColor.fgDim)
        case .available:
            Image(systemName: "checkmark.circle.fill")
                .font(.system(size: 12, weight: .semibold))
                .foregroundStyle(TaliseColor.accent)
        case .taken, .reserved, .invalid:
            Image(systemName: "exclamationmark.circle")
                .font(.system(size: 12, weight: .regular))
                .foregroundStyle(TaliseColor.danger)
        case .empty, .rpcError:
            Color.clear.frame(width: 12, height: 12)
        }
    }

    private var statusText: String {
        switch availability {
        case .empty:     return ""
        case .checking:  return "Checking…"
        case .available: return "\(input)@talise.sui is available."
        case .taken:     return "Someone already claimed that name."
        case .reserved:  return "That name is reserved."
        case .invalid:   return "Use 3–20 lowercase letters, digits, or underscores."
        case .rpcError:  return "Couldn't check on chain. Tap claim anyway."
        }
    }

    private var statusColor: Color {
        switch availability {
        case .available: return TaliseColor.accent
        case .taken, .reserved, .invalid: return TaliseColor.danger
        case .rpcError: return TaliseColor.fgMuted
        default: return TaliseColor.fgDim
        }
    }

    private var claimButton: some View {
        Button {
            Task { await claim() }
        } label: {
            HStack(spacing: 8) {
                if claiming {
                    ProgressView().controlSize(.small).tint(TaliseColor.bg)
                } else {
                    Image(systemName: "checkmark.seal.fill")
                        .font(.system(size: 13, weight: .medium))
                }
                Text(claiming ? "Claiming…" : "Claim \(input)@talise.sui")
                    .font(TaliseFont.heading(15, weight: .medium))
                    .lineLimit(1)
                    .minimumScaleFactor(0.7)
            }
            .foregroundStyle(TaliseColor.bg)
            .frame(maxWidth: .infinity)
            .frame(height: 52)
            .background(canClaim ? TaliseColor.fg : TaliseColor.fg.opacity(0.35))
            .clipShape(Capsule())
        }
        .disabled(!canClaim)
        .buttonStyle(.plain)
    }

    private var canClaim: Bool {
        !claiming && (availability == .available || availability == .rpcError)
    }

    // MARK: - Success

    private func successView(handle: String) -> some View {
        VStack(spacing: 16) {
            Spacer()
            ZStack {
                Circle().fill(TaliseColor.accent.opacity(0.15)).frame(width: 84, height: 84)
                Image(systemName: "checkmark.seal.fill")
                    .font(.system(size: 32, weight: .semibold))
                    .foregroundStyle(TaliseColor.accent)
            }
            Text("Claimed")
                .font(TaliseFont.heading(28, weight: .medium))
                .kerning(-1)
                .foregroundStyle(TaliseColor.fg)
            Text("\(handle)@talise.sui is yours.")
                .font(TaliseFont.body(14, weight: .light))
                .foregroundStyle(TaliseColor.fgMuted)
            Spacer()
            Button {
                dismiss()
                Task { await session.bootstrap() }
            } label: {
                Text("Done")
                    .font(TaliseFont.heading(15, weight: .medium))
                    .foregroundStyle(TaliseColor.bg)
                    .frame(maxWidth: .infinity)
                    .frame(height: 52)
                    .background(TaliseColor.fg)
                    .clipShape(Capsule())
            }
            .padding(.horizontal, 24)
            .padding(.bottom, 40)
        }
    }

    // MARK: - Helpers

    private func sanitize(_ s: String) -> String {
        // Server-side normalizeHandle accepts the same character set:
        // lowercased [a-z0-9_], 3-20 chars. We mirror it client-side
        // so the input box reflects what'll actually be sent.
        let allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyz0123456789_")
        let cleaned = s.lowercased().unicodeScalars
            .filter { allowed.contains($0) }
            .map(String.init).joined()
        return String(cleaned.prefix(20))
    }

    private func scheduleCheck(_ raw: String) {
        checkTask?.cancel()
        error = nil
        let q = raw.trimmingCharacters(in: .whitespaces)
        if q.isEmpty {
            availability = .empty
            return
        }
        if q.count < 3 {
            availability = .invalid
            return
        }
        availability = .checking
        checkTask = Task {
            try? await Task.sleep(nanoseconds: 250_000_000)
            if Task.isCancelled { return }
            do {
                let encoded = q.addingPercentEncoding(
                    withAllowedCharacters: .urlQueryAllowed
                ) ?? q
                let r: UsernameCheckResponse = try await APIClient.shared.get(
                    "/api/username/check?u=\(encoded)"
                )
                if Task.isCancelled { return }
                if r.available {
                    availability = .available
                } else {
                    switch r.reason {
                    case "taken":    availability = .taken
                    case "reserved": availability = .reserved
                    case "invalid":  availability = .invalid
                    case "rpc":      availability = .rpcError
                    default:         availability = .invalid
                    }
                }
            } catch {
                if Task.isCancelled { return }
                availability = .rpcError
            }
        }
    }

    private func claim() async {
        claiming = true
        error = nil
        defer { claiming = false }
        struct Body: Encodable { let username: String }
        do {
            let _: UsernameClaimResponse = try await APIClient.shared.post(
                "/api/username/claim",
                body: Body(username: input)
            )
            claimed = input
        } catch APIError.status(let code, let msg) where code == 409 {
            error = msg ?? "That name was just taken."
            availability = .taken
        } catch APIError.status(_, let msg) {
            error = msg ?? "Couldn't claim that handle right now."
        } catch let claimErr {
            error = claimErr.localizedDescription
        }
    }
}

File: android/app/src/main/java/io/talise/app/feature/profile/ClaimHandleSheet.kt
package io.talise.app.feature.profile

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.ErrorOutline
import androidx.compose.material.icons.filled.Verified
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.talise.app.core.model.UserDTO
import io.talise.app.ui.components.LiquidGlassButton
import io.talise.app.ui.components.MicroLabel
import io.talise.app.ui.theme.TaliseColors
import io.talise.app.ui.theme.TaliseType
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Bottom sheet that lets the user claim a `<name>.talise.sui` subname — exact port
 * of iOS `ClaimHandleSheet`.
 *
 * Flow:
 *   1. Debounced GET /api/username/check?u=<input> on every keystroke
 *   2. Tap "Claim" → POST /api/username/claim — operator wallet pays gas + signs
 *      the SuiNS mint, user pays nothing
 *   3. On success, refresh the session user so Home/Profile pick up the new handle
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ClaimHandleSheet(user: UserDTO?, onClaimed: () -> Unit, onDismiss: () -> Unit) {
    var input by remember { mutableStateOf("") }
    var availability by remember { mutableStateOf(HandleAvailability.EMPTY) }
    var claiming by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    var claimed by remember { mutableStateOf<String?>(null) }
    var checkJob by remember { mutableStateOf<Job?>(null) }
    val scope = rememberCoroutineScope()
    val focusRequester = remember { FocusRequester() }

    fun scheduleCheck(raw: String) {
        checkJob?.cancel()
        error = null
        val q = raw.trim()
        if (q.isEmpty()) {
            availability = HandleAvailability.EMPTY
            return
        }
        if (q.length < 3) {
            availability = HandleAvailability.INVALID
            return
        }
        availability = HandleAvailability.CHECKING
        checkJob = scope.launch {
            delay(250)
            try {
                val r = profileApi.usernameCheck(q)
                availability = if (r.available) {
                    HandleAvailability.AVAILABLE
                } else {
                    when (r.reason) {
                        "taken" -> HandleAvailability.TAKEN
                        "reserved" -> HandleAvailability.RESERVED
                        "invalid" -> HandleAvailability.INVALID
                        "rpc" -> HandleAvailability.RPC_ERROR
                        else -> HandleAvailability.INVALID
                    }
                }
            } catch (e: CancellationException) {
                throw e
            } catch (_: Exception) {
                availability = HandleAvailability.RPC_ERROR
            }
        }
    }

    fun claim() {
        if (claiming) return
        claiming = true
        error = null
        scope.launch {
            try {
                profileApi.usernameClaim(UsernameClaimBody(input))
                claimed = input
            } catch (t: Throwable) {
                val code = httpCode(t)
                val msg = httpErrorMessage(t)
                if (code == 409) {
                    error = msg ?: "That name was just taken."
                    availability = HandleAvailability.TAKEN
                } else if (code != null) {
                    error = msg ?: "Couldn't claim that handle right now."
                } else {
                    error = t.message ?: "Couldn't claim that handle right now."
                }
            }
            claiming = false
        }
    }

    LaunchedEffect(Unit) {
        if (input.isEmpty()) {
            input = suggestedHandle(user)
            scheduleCheck(input)
            focusRequester.requestFocus()
        }
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
        containerColor = TaliseColors.bg,
    ) {
        val done = claimed
        if (done != null) {
            ClaimSuccessView(
                handle = done,
                onDone = {
                    onDismiss()
                    onClaimed()
                },
            )
        } else {
            Column(
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 24.dp)
                    .padding(top = 8.dp),
                verticalArrangement = Arrangement.spacedBy(22.dp),
            ) {
                Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                    MicroLabel("Claim your name", color = TaliseColors.fgDim)
                    Text(
                        "Pick your Talise handle",
                        style = TaliseType.heading(24.sp, FontWeight.Medium),
                        letterSpacing = (-0.8).sp,
                        color = TaliseColors.fg,
                    )
                    Text(
                        "People send to you with name@talise.sui, easier to share than a 0x address.",
                        style = TaliseType.body(13.sp, FontWeight.Light),
                        color = TaliseColors.fgMuted,
                        modifier = Modifier.padding(top = 2.dp),
                    )
                }

                // Input — sanitized live to the server's normalizeHandle char set.
                Row(
                    Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(16.dp))
                        .background(
                            if (availability == HandleAvailability.AVAILABLE)
                                TaliseColors.accent.copy(alpha = 0.12f)
                            else TaliseColors.surface,
                            RoundedCornerShape(16.dp),
                        )
                        .padding(horizontal = 16.dp, vertical = 16.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Box(Modifier.weight(1f)) {
                        if (input.isEmpty()) {
                            Text(
                                "alice",
                                style = TaliseType.heading(20.sp, FontWeight.Medium),
                                letterSpacing = (-0.4).sp,
                                color = TaliseColors.fgDim,
                            )
                        }
                        BasicTextField(
                            value = input,
                            onValueChange = { new ->
                                val cleaned = sanitizeHandle(new)
                                input = cleaned
                                scheduleCheck(cleaned)
                            },
                            singleLine = true,
                            textStyle = TaliseType.heading(20.sp, FontWeight.Medium)
                                .copy(color = TaliseColors.fg, letterSpacing = (-0.4).sp),
                            cursorBrush = SolidColor(TaliseColors.accent),
                            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Ascii),
                            modifier = Modifier
                                .fillMaxWidth()
                                .focusRequester(focusRequester),
                        )
                    }
                    Text(
                        "@talise.sui",
                        style = TaliseType.body(15.sp, FontWeight.Light),
                        color = TaliseColors.fgMuted,
                    )
                }

                // Status row
                Row(
                    Modifier.height(18.dp).padding(horizontal = 4.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    when (availability) {
                        HandleAvailability.CHECKING ->
                            CircularProgressIndicator(
                                color = TaliseColors.fgDim,
                                strokeWidth = 1.5.dp,
                                modifier = Modifier.size(12.dp),
                            )
                        HandleAvailability.AVAILABLE ->
                            Icon(
                                Icons.Filled.CheckCircle,
                                contentDescription = null,
                                tint = TaliseColors.accent,
                                modifier = Modifier.size(12.dp),
                            )
                        HandleAvailability.TAKEN,
                        HandleAvailability.RESERVED,
                        HandleAvailability.INVALID ->
                            Icon(
                                Icons.Filled.ErrorOutline,
                                contentDescription = null,
                                tint = TaliseColors.danger,
                                modifier = Modifier.size(12.dp),
                            )
                        HandleAvailability.EMPTY, HandleAvailability.RPC_ERROR ->
                            Box(Modifier.size(12.dp))
                    }
                    Text(
                        statusText(availability, input),
                        style = TaliseType.body(12.sp, FontWeight.Light),
                        color = statusColor(availability),
                    )
                }

                error?.let { MicroLabel(it, color = TaliseColors.danger) }

                val canClaim = !claiming &&
                    (availability == HandleAvailability.AVAILABLE || availability == HandleAvailability.RPC_ERROR)
                LiquidGlassButton(
                    title = if (claiming) "Claiming…" else "Claim $input@talise.sui",
                    onClick = { claim() },
                    tint = if (canClaim) TaliseColors.greenMint else null,
                    enabled = canClaim,
                    loading = claiming,
                )

                Spacer(Modifier.height(40.dp))
            }
        }
    }
}

// MARK: - Success

@Composable
private fun ClaimSuccessView(handle: String, onDone: () -> Unit) {
    Column(
        Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        Spacer(Modifier.height(24.dp))
        Box(
            Modifier
                .size(84.dp)
                .clip(CircleShape)
                .background(TaliseColors.accent.copy(alpha = 0.16f)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                Icons.Filled.Verified,
                contentDescription = null,
                tint = TaliseColors.accent,
                modifier = Modifier.size(32.dp),
            )
        }
        Text(
            "Claimed",
            style = TaliseType.heading(28.sp, FontWeight.Medium),
            letterSpacing = (-1).sp,
            color = TaliseColors.fg,
        )
        Text(
            "$handle@talise.sui is yours.",
            style = TaliseType.body(14.sp, FontWeight.Light),
            color = TaliseColors.fgMuted,
        )
        Spacer(Modifier.height(24.dp))
        LiquidGlassButton(
            title = "Done",
            onClick = onDone,
            tint = TaliseColors.greenMint,
            modifier = Modifier.padding(horizontal = 24.dp),
        )
        Spacer(Modifier.height(40.dp))
    }
}

// MARK: - State helpers

internal enum class HandleAvailability {
    EMPTY, CHECKING, AVAILABLE, TAKEN, RESERVED, INVALID, RPC_ERROR
}

private fun statusText(a: HandleAvailability, input: String): String = when (a) {
    HandleAvailability.EMPTY -> ""
    HandleAvailability.CHECKING -> "Checking…"
    HandleAvailability.AVAILABLE -> "$input@talise.sui is available."
    HandleAvailability.TAKEN -> "Someone already claimed that name."
    HandleAvailability.RESERVED -> "That name is reserved."
    HandleAvailability.INVALID -> "Use 3-20 lowercase letters, digits, or underscores."
    HandleAvailability.RPC_ERROR -> "Couldn't check on chain. Tap claim anyway."
}

@Composable
private fun statusColor(a: HandleAvailability) = when (a) {
    HandleAvailability.AVAILABLE -> TaliseColors.accent
    HandleAvailability.TAKEN, HandleAvailability.RESERVED, HandleAvailability.INVALID -> TaliseColors.danger
    HandleAvailability.RPC_ERROR -> TaliseColors.fgMuted
    else -> TaliseColors.fgDim
}

/**
 * Server-side normalizeHandle accepts the same character set: lowercased
 * [a-z0-9_], 3-20 chars. Mirrored client-side so the input box reflects
 * what'll actually be sent.
 */
private fun sanitizeHandle(s: String): String =
    s.lowercase().filter { it in 'a'..'z' || it in '0'..'9' || it == '_' }.take(20)

/**
 * Suggestion used to seed the claim sheet — derived from the Google name (then
 * email local-part). Never shown standalone as if it were the user's real handle.
 */
private fun suggestedHandle(user: UserDTO?): String {
    val source = user?.name?.takeIf { it.isNotEmpty() }?.split(" ")?.firstOrNull()
        ?: user?.email?.substringBefore("@")
        ?: "you"
    return sanitizeHandle(source)
}
